### PR TITLE
Add pandas index checks

### PR DESCRIPTION
**Pull request by skrawcz**
Adds some basic index type checking to df creation
    
Related to issue #191, this commit is to help surface index type issues.
    
Specifically:
    
1. Warn if there are index type mismatches.
2. Require you to set your logger to debug if you want to see more details.
3. Provide a "ResultBuilder" class that uses strict index type matching so if you
want to error on index type mismatches, this is the results builder to use.
    
I don't think we should build anything more custom unless there's a clear
common use case - user contributed result builders sound like an interesting idea here though.

To use the new result builder the code should be the following:

```python
from hamilton import base, driver
strict_builder = base.StrictIndexTypePandasDataFrameResult()
adapter = base.SimplePythonGraphAdapter(strict_builder)
...
dr =  driver.Driver(config, *modules, adapter=adapter)
df = dr.execute(...)  # this will now error if index types mismatch.
```

## Changes

- one clean up commit around imports
- one commit with the index type changes
- one commit to handle python 3.6 support and pandas index classes changing

## Testing

1. Unit tests.
2. Checked things in the REPL locally against the examples in #191.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python - local testing

- [ ] python 3.6
- [ ] python 3.7
- [ ] python 3.8
- [x] python 3.9

### CODE REVIEW

**Review by elijahbenizzy — COMMENTED** [review]
Overall looks good -- curious on some decisions though around time-series

File: hamilton/base.py [hunk 1]
```diff
@@ -41,9 +45,99 @@ class PandasDataFrameResult(ResultMixin):
...
+    ) -> Tuple[Dict[str, List[str]], Dict[str, List[str]], Dict[str, List[str]]]:
+        """This function creates three dictionaries according to whether there is an index type or not.
+
+        The three dicts we create are:
+        1. Dict of index type to list of outputs that match it.
+        2. Dict of time series / categorical index types to list of outputs that match it.
+        3. Dict of `no-index` key to list of outputs with no index type.
+
+        :param outputs: the dict we're trying to create a result from.
+        :return: dict of all index types, dict of time series/categorical index types, dict if there is no index
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py:59]: That's the inheritance structure of pandas indexes.

File: hamilton/base.py [hunk 2]
```diff
@@ -41,9 +45,99 @@ class PandasDataFrameResult(ResultMixin):
...
+        """
+        all_index_types = collections.defaultdict(list)
+        time_indexes = collections.defaultdict(list)
+        no_indexes = collections.defaultdict(list)
+        for output_name, output_value in outputs.items():
+            if isinstance(output_value, (pd.DataFrame, pd.Series)):
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py:65]: sure - I can handle that.

File: hamilton/base.py [hunk 3]
```diff
@@ -41,9 +45,99 @@ class PandasDataFrameResult(ResultMixin):
...
+                dict_key = f"{output_value.index.__class__.__name__}:::{output_value.index.dtype}"
+                try:
+                    index_type = getattr(pd_extension, "NDArrayBackedExtensionIndex")
+                except AttributeError:  # for python 3.6 & pandas 1.1.5
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py:69]: yup --

File: hamilton/base.py [hunk 4]
```diff
@@ -41,9 +45,99 @@ class PandasDataFrameResult(ResultMixin):
...
+
+        This only logs warning errors, and if debug is enabled, a debug statement to list index types.
+        """
+        no_index_length = len(no_indexes)
+        time_indexes_length = len(time_indexes)
+        all_indexes_length = len(all_index_types)
+        number_with_indexes = all_indexes_length - no_index_length
+        types_match = True  # default to True
+        # if there is more than one time index
+        if time_indexes_length > 1:
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py:100]: yep -- only index type check.

File: hamilton/base.py [hunk 5]
```diff
@@ -54,14 +148,45 @@ def build_result(**outputs: typing.Dict[str, typing.Any]) -> pd.DataFrame:
...
+    def build_result(**outputs: Dict[str, Any]) -> pd.DataFrame:
+        # TODO check inputs are pd.Series, arrays, or scalars -- else error
+        output_index_type_tuple = PandasDataFrameResult.pandas_index_types(outputs)
+        indexes_match = PandasDataFrameResult.check_pandas_index_types_match(
+            *output_index_type_tuple
+        )
+        if not indexes_match:
+            raise ValueError(
+                "Error: pandas index types did not match exactly. "
+                f"Found the following indexes:\n{dict(output_index_type_tuple[0])}"
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py:180]: sure.

File: hamilton/base.py [hunk 6]
```diff
@@ -41,9 +45,99 @@ class PandasDataFrameResult(ResultMixin):
...
+        :param outputs: the dict we're trying to create a result from.
+        :return: dict of all index types, dict of time series/categorical index types, dict if there is no index
+        """
+        all_index_types = collections.defaultdict(list)
+        time_indexes = collections.defaultdict(list)
+        no_indexes = collections.defaultdict(list)
+        for output_name, output_value in outputs.items():
+            if isinstance(output_value, (pd.DataFrame, pd.Series)):
+                dict_key = f"{output_value.index.__class__.__name__}:::{output_value.index.dtype}"
+                try:
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py:67]: hmm not sure I can make this any cleaner..

[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/base.py]: ok. done.

File: hamilton/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply hamilton/base.py:119]: I'm not sure this is of that much value, as by default we'll expect that there will e multiple time indices. So then they'll always get the warning message. Unless I'm misreading it? I'd prefer we change the warning message at least.

File: hamilton/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply hamilton/base.py:59]: Feels like we shouldn't index to that and we should be able to differentiate.

**Review by elijahbenizzy — COMMENTED** [review]
Yeah, this is all good except the time-series stuff is confusing me.

**Review by elijahbenizzy — APPROVED** [review]

**Pull request merged**